README.md
[![Supported Python versions](https://img.shields.io/pypi/pyversions/coffeepy.svg?style=flat)](https://pypi.python.org/pypi/coffeepy/) [![PyPI Version](https://img.shields.io/pypi/v/coffeepy.svg)](https://pypi.python.org/pypi/coffeepy)
# Coffeepy

Coffeepy ☕️ is a small program that prevents the system from sleeping.
Works on MacOS, Windows and Linux.

<img alt="coffeepy" src="https://github-production-user-asset-6210df.s3.amazonaws.com/22169537/253075028-9eaccaca-a567-4bd8-86c1-63d4870664ad.gif" width="600" />

## Installation

```sh
pip install -U coffeepy
```

## Usage

Simply run the program from command line
```sh
coffeepy
```

By default the program runs indefinitely. Press `Ctrl-C` to quit

Optional: You can set the time in minutes with `-t` or `--time` flag. For example, to run the program for 1 hour, use

```sh
coffeepy -t 60
```
You can also disable animation with `-a` or `--no-animation` flag.
```sh
coffeepy -a
```
You can view the full parameter list with `-h` or `--help`.

## Python module

You can also import coffeepy as a python module

```python
import coffeepy

coffeepy.run()
```

Optional settings when run as a python module:

```python
import coffeepy

# you can also specify the time in minutes
# if no time is provided or time = 0, the program will run indefinitely
coffeepy.run(60)

# to disable animation when run as a module, you can set the second argument to True
coffeepy.run(0, True)
```


## License

This project is licensed under the MIT [License](https://github.com/kuvaus/coffeepy/blob/main/LICENSE)


CHANGELOG.md

## Changelog

#### [Upcoming](https://github.com/kuvaus/coffeepy/compare/v0.2.1...HEAD)

#### [v0.2.1](https://github.com/kuvaus/coffeepy/releases/tag/v0.2.1)

> 24 July 2023

- Add ability to reset the system idle timer on Windows
- This prevents computer from going to sleep even if no other programs are running

#### [v0.2.0](https://github.com/kuvaus/coffeepy/releases/tag/v0.2.0)

> 23 July 2023

- Simplify dependencies

#### [v0.1.9](https://github.com/kuvaus/coffeepy/releases/tag/v0.1.9)

> 23 July 2023

- Finally works properly on various linux systems
- Use `DBUS` on linux to prevent system from sleeping  [`#1`](https://github.com/kuvaus/coffeepy/pull/1)
- Use `systemd` on linux as a fallback method
- Deprecate old linux method of using `xset`
- Automatic release notes
- Improve tests

#### [v0.1.8](https://github.com/kuvaus/coffeepy/releases/tag/v0.1.8)

> 22 July 2023

- Fix `--help` on windows systems without unicode support
- Better error handling on Linux
- Add Python 3.7 support

#### [v0.1.7](https://github.com/kuvaus/coffeepy/releases/tag/v0.1.7)

> 16 July 2023

- Significantly improved performance when animation is turned off
- Improve tests

#### [v0.1.6](https://github.com/kuvaus/coffeepy/releases/tag/v0.1.6)

> 14 July 2023

- Corrected Python requirements to 3.8 and newer
- Time can now be a float so you can run it more accurately
- Add tests. Run the tests with `pytest`

#### [v0.1.5](https://github.com/kuvaus/coffeepy/releases/tag/v0.1.5)

> 13 July 2023

- Improve terminal checking function

#### [v0.1.4](https://github.com/kuvaus/coffeepy/releases/tag/v0.1.4)

> 13 July 2023

- Fix bug in checking new Windows Terminal
- Add ability to specify time=0 when run as a module
- Add ability to disable animation when run as a module

#### [v0.1.3](https://github.com/kuvaus/coffeepy/releases/tag/v0.1.3)

> 13 July 2023

- Add `--no-animation` flag
- Add ascii animation on Windows Powershell
- Fix duration bug from 0.1.2

#### [v0.1.2](https://github.com/kuvaus/coffeepy/releases/tag/v0.1.2)

> 13 July 2023

- Ability to set duration when run as a python module
- Add .whl and .tar.gz to Github releases

#### [v0.1.1](https://github.com/kuvaus/coffeepy/releases/tag/v0.1.1)

> 12 July 2023

- Automatic Pypi distribution with Github actions
- Better project structure

#### v0.1.0

> 12 July 2023

- First version


tests/test_coffeepy.py
import pytest
import sys
from unittest.mock import patch, call, Mock, MagicMock
from io import StringIO
import subprocess
import builtins
from coffeepy import *

#
# Tests for:
#
# animation, ascii_animation
# display_animation()
# check_caffeinate()
# check_windows_terminal()
# parse_args()
# run()
#


#
# Test the display_animation function
#

# see if the function runs
def test_display_animation():
    display_animation()
    display_animation(animation)
    display_animation(ascii_animation)

# check with test_animation
@patch('builtins.print')
def test_display_animation_test_animation(mock_print):
    test_animation = [
        "  ☕️   ",
        " ☁️☕️   ",
        "  ☕️   ",
    ]
    display_animation(animation=test_animation)
    calls = [call('\r' + frame, end='') for frame in test_animation]
    mock_print.assert_has_calls(calls)


# check that the default animation prints the same as animation
@patch('builtins.print')
def test_display_animation_default_animation(mock_print):
    display_animation()
    calls = [call('\r' + frame, end='') for frame in animation]
    mock_print.assert_has_calls(calls)

#
# Test the check_caffeinate function
# True if caffeinate is found on /usr/bin
#

@patch('subprocess.check_output')
def test_check_caffeinate(mock_subproc):
    mock_subproc.return_value = '/usr/bin/caffeinate'
    assert check_caffeinate() is True

    mock_subproc.side_effect = subprocess.CalledProcessError(1, 'which')
    assert check_caffeinate() is False


#
# Test the check_windows_terminal function
# True if WT_SESSION is found on win32, false otherwise
#

# windows
@patch('os.environ.get')
@patch('sys.platform', 'win32')
def test_check_windows_terminal_win(mock_get):
    mock_get.return_value = None
    assert check_windows_terminal() is False

    mock_get.return_value = 'WT_SESSION'
    assert check_windows_terminal() is True

# linux
@patch('os.environ.get')
@patch('sys.platform', 'linux')
def test_check_windows_terminal_linux(mock_get):
    mock_get.return_value = None
    assert check_windows_terminal() is False

    mock_get.return_value = 'WT_SESSION'
    assert check_windows_terminal() is False

# macos
@patch('os.environ.get')
@patch('sys.platform', 'darwin')
def test_check_windows_terminal_macos(mock_get):
    mock_get.return_value = None
    assert check_windows_terminal() is False

    mock_get.return_value = 'WT_SESSION'
    assert check_windows_terminal() is False

#
# Test argument parsing
#

def test_argument_parsing_default():
    args = parse_args([])
    assert args.time == 0
    assert args.no_animation is False

def test_argument_parsing_time():
    args = parse_args(['-t', '10'])
    assert args.time == 10

def test_argument_parsing_no_animation():
    args = parse_args(['-a'])
    assert args.no_animation


#
# Test the run function
# First make sure this is run on a system with caffeinate
#

if check_caffeinate():
    def test_timed_run():
        runtime = 0.01
        run(runtime)

    def test_timed_run():
        runtime = 0.01
        run(runtime)

#
# Test run on macos, linux and windows
#

# macOS
# Use mock to simulate 'check_caffeinate'
#
@patch('sys.platform', new='darwin')
@patch('subprocess.Popen')
@patch('subprocess.check_output')
def test_platform_macos(mock_subproc, mock_popen):
    mock_subproc.return_value = '/usr/bin/caffeinate'
    mock_popen.return_value.returncode = 0
    runtime = 0.01
    run(runtime)
    mock_popen.assert_called_once_with(['caffeinate', '-dims'])

#
# Linux and caffeinate
# Use mock to simulate 'check_caffeinate'
#
@patch('sys.platform', new='linux')
@patch('subprocess.Popen')
@patch('subprocess.check_output')
def test_platform_linux_with_caffeinate(mock_subproc, mock_popen):
    mock_subproc.return_value = '/usr/bin/caffeinate'
    mock_popen.return_value.returncode = 0
    runtime = 0.01
    run(runtime)
    mock_popen.assert_called_once_with(['caffeinate', '-dims'])


#
# Windows
# Use mock to simulate 'ctypes.windll.kernel32.SetThreadExecutionState'
#
@patch('sys.platform', new='win32')
@patch('ctypes.windll', create=True)
def test_platform_windows(self):
    runtime = 0.01
    run(runtime)



coffeepy/main.py
import ctypes
import sys
import subprocess
import time
import argparse


animation = [
    "  ☕️   ",
    " ☁️☕️   ",
    "  ☕️   ",
    "  ☕️☁️  ",
    " ☁️☕️   ",
    "  ☕️   ",
]


def display_animation(animation):
    for frame in animation:
        print('\r' + frame, end='')
        time.sleep(0.5)  # Adjust the delay time as desired


def check_caffeinate():
    try:
        subprocess.check_output(['which', 'caffeinate'])
        return True
    except subprocess.CalledProcessError:
        return False


def main():
    parser = argparse.ArgumentParser(description='Coffeepy ☕️ prevents the system from sleeping.\n'
                                                 'You can set the time with -t flag\n'
                                                 'Made by kuvaus',
                                                 formatter_class=argparse.RawTextHelpFormatter)

    parser.add_argument('-t', '--time', type=int, default=0, help='Optional: Duration of animation in minutes. Use 0 for indefinite duration')
    args = parser.parse_args()

    if args.time == 0:
        duration = float('inf')  # Set duration to infinity
    else:
        duration = args.time * 60  # Convert minutes to seconds
        
    proc = None

    if 'darwin' in sys.platform:
        print('Running \'coffeepy\' on MacOS to prevent the system from sleeping')
        proc = subprocess.Popen(['caffeinate', '-dims'])

    elif 'linux' in sys.platform:
        print('Running \'coffeepy\' on Linux to prevent the system from sleeping')
        if check_caffeinate():
            proc = subprocess.Popen(['caffeinate', '-dims'])
        else:
            subprocess.Popen(['xset', 's', 'off'])
            subprocess.Popen(['xset', '-dpms'])

    elif 'win32' in sys.platform:
        print('Running \'coffeepy\' on Windows to prevent the system from sleeping')
        ctypes.windll.kernel32.SetThreadExecutionState(0x80000002)

    print('Press Ctrl-C to quit')
    
    try:
        start_time = time.time()
        while time.time() - start_time < duration or duration == float('inf'):
            display_animation(animation)
    except KeyboardInterrupt:
        print('\nExiting')

    finally:
        if proc:
            proc.terminate()
        if 'linux' in sys.platform and not check_caffeinate():
            # Reset xset settings
            subprocess.Popen(['xset', 's', 'on'])
            subprocess.Popen(['xset', '+dpms'])
        if 'win32' in sys.platform:
            ctypes.windll.kernel32.SetThreadExecutionState(0x80000000)
        sys.exit()


if __name__ == "__main__":
    main()


src/coffeepy/coffeepy.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import ctypes
import sys
import subprocess
import time
import argparse
import os
import warnings

if 'linux' in sys.platform:
    import jeepney
    from jeepney.io.blocking import open_dbus_connection

animation = [
    "  ☕️   ",
    " ☁️☕️   ",
    "  ☕️   ",
    "  ☕️☁️  ",
    " ☁️☕️   ",
    "  ☕️   ",
]

ascii_animation = [
    "  |   ",
    "  /   ",
    "  -   ",
    "  \\   ",
]

def display_animation(animation=animation):
    for frame in animation:
        print('\r' + frame, end='')
        time.sleep(0.5)  # Adjust the delay time as desired

def get_version(package):
    if sys.version_info >= (3, 8):
        from importlib.metadata import version
        return version(package)
    else:
        with warnings.catch_warnings():
            warnings.filterwarnings("ignore", category=DeprecationWarning)
            import pkg_resources
            return pkg_resources.get_distribution(package).version


# On macOS check caffeinate
def check_caffeinate():
    try:
        subprocess.check_output(['which', 'caffeinate'])
        return True
    except subprocess.CalledProcessError:
        return False

# On Windows check if we use Windows Terminal
def check_windows_terminal():
    if ('win32' in sys.platform) and (os.environ.get("WT_SESSION") is not None):
        return True
    else:
        return False

# On Linux use DBUS
def get_connection():
    try:
        connection = open_dbus_connection(bus="SESSION")
    except Exception as e:
        print("Could not set DBUS SESSION")
        connection = None
    return connection

def set_dbus_awake(connection):
    proxy = jeepney.DBusAddress('/org/freedesktop/ScreenSaver', bus_name='org.freedesktop.ScreenSaver', interface='org.freedesktop.ScreenSaver')
    msg = jeepney.new_method_call(proxy, "Inhibit", "ss", ("coffeepy", "wakelock active"))
    reply = connection.send_and_get_reply(msg)
    cookie = reply.body[0]
    return cookie

def unset_dbus_awake(connection, cookie):
    if cookie is None:
        return
    else:
        proxy = jeepney.DBusAddress('/org/freedesktop/ScreenSaver', bus_name='org.freedesktop.ScreenSaver', interface='org.freedesktop.ScreenSaver')
        msg = jeepney.new_method_call(proxy, "UnInhibit", "u", (cookie,))
        connection.send_and_get_reply(msg)

# On Linux use systemd if DBUS fails
def set_systemd_mask():
    print("Trying with systemd (this will need sudo permissions)")
    result = subprocess.run(["systemctl", "mask", "sleep.target", "suspend.target", "hibernate.target", "hybrid-sleep.target"])
    if result.returncode != 0:
        print("Systemd failed.")
        sys.exit(0)
        
def unset_systemd_mask():
    result = subprocess.run(["systemctl", "unmask", "sleep.target", "suspend.target", "hibernate.target", "hybrid-sleep.target"])
    if result.returncode != 0:
        print("Systemd failed.")
        sys.exit(0)

# Parse CLI arguments
def parse_args(args=None):

    coffee_emoji = "☕️"
    if 'win32' in sys.platform and not check_windows_terminal():
        coffee_emoji = ""

    description = 'Coffeepy (v'+get_version('coffeepy')+') '+coffee_emoji+""" prevents the system from sleeping.
You can set the time with -t flag
Made by kuvaus"""

    parser = argparse.ArgumentParser(description=description, formatter_class=argparse.RawTextHelpFormatter)

    parser.add_argument('-t', '--time', type=float, default=0, help='Optional: Duration of run in minutes. Use 0 for infinite duration')
    parser.add_argument('-a', '--no-animation', action='store_true', help='Optional: Disable animation')

    return parser.parse_args(args)
    
def run(runtime=0, no_animation=False):

    args = None
    if runtime > 0 or no_animation:
        args = ['-t', str(runtime)]
        if no_animation:
            args.append('-a')

    args = parse_args(args)
        
    if args.time == 0:
        duration = float('inf')  # Set duration to infinity
    else:
        duration = args.time * 60  # Convert minutes to seconds

    proc = None
    connection = None

    if 'darwin' in sys.platform:
        print('Running \'coffeepy\' on MacOS to prevent the system from sleeping')
        proc = subprocess.Popen(['caffeinate', '-dims'])

    elif 'linux' in sys.platform:
        print('Running \'coffeepy\' on Linux to prevent the system from sleeping')
        if check_caffeinate():
            proc = subprocess.Popen(['caffeinate', '-dims'])
        else:
            # Check for DBUS
            connection = get_connection()
            if connection is not None:
            # Keep system awake using DBUS
                cookie = set_dbus_awake(connection)
            else:
            # Keep system awake using systemd
                print("You need to install either \'dbus\' or \'caffeinate\' package for this program to run")
                set_systemd_mask()


    elif 'win32' in sys.platform:
        print('Running \'coffeepy\' on Windows to prevent the system from sleeping')
        ctypes.windll.kernel32.SetThreadExecutionState(0x80000003)

    print('Press Ctrl-C to quit')
    
    try:
        start_time = time.time()
        while time.time() - start_time < duration or duration == float('inf'):
            if not args.no_animation:
                if 'win32' in sys.platform and not check_windows_terminal():
                    display_animation(ascii_animation)
                else:
                    display_animation()
            else:
                time.sleep(1)

    except KeyboardInterrupt:
        print('\nExiting')

    finally:
        if proc:
            proc.terminate()
        if 'linux' in sys.platform and not check_caffeinate():
            if connection is not None:
                # Reset DBUS connection
                unset_dbus_awake(connection, cookie)
            else:
                # Reset systemd changes
                unset_systemd_mask()
        if 'win32' in sys.platform:
            ctypes.windll.kernel32.SetThreadExecutionState(0x80000000)


if __name__ == "__main__":
    run()
